src/mito2/src/read/merge.rs

```diff
@@ -176,36 +181,58 @@ impl MergeReader {
-                // must be less than `next_min_ts`.
-                Self::maybe_output_batch(
-                    top.slice(0, pos),
-                    &mut self.output_batch,
-                    &mut self.metrics,
-                )?;
-                // This keep the duplicate timestamp in the node.
-                top_node.skip_rows(pos, &mut self.metrics).await?;
-                // The merge window should contain this timestamp so only nodes in the hot heap
-                // have this timestamp.
-                self.filter_first_duplicate_timestamp_in_hot(top_node, next_min_ts)
-                    .await?;
-            }
+        let duplicate_pos = match timestamps.binary_search(&next_min_ts.value()) {
+            Ok(pos) => pos,
             Err(pos) => {
                 // No duplicate timestamp. Outputs timestamp before `pos`.
                 Self::maybe_output_batch(
                     top.slice(0, pos),
                     &mut self.output_batch,
+                    self.dedup,
                     &mut self.metrics,
                 )?;
                 top_node.skip_rows(pos, &mut self.metrics).await?;
-                self.reheap(top_node)?;
+                return self.reheap(top_node);
             }
+        };
+
+        if self.dedup {
+            // They have duplicate timestamps. Outputs timestamps before the duplicated timestamp.
+            // Batch itself doesn't contain duplicate timestamps so timestamps before `duplicate_pos`
+            // must be less than `next_min_ts`.
+            Self::maybe_output_batch(
+                top.slice(0, duplicate_pos),
+                &mut self.output_batch,
+                self.dedup,
+                &mut self.metrics,
+            )?;
+            // This keep the duplicate timestamp in the node.
+            top_node.skip_rows(duplicate_pos, &mut self.metrics).await?;
+            // The merge window should contain this timestamp so only nodes in the hot heap
+            // have this timestamp.
+            return self
+                .filter_first_duplicate_timestamp_in_hot(top_node, next_min_ts)
+                .await;
         }
 
-        Ok(())
+        // No need to remove duplicate timestamps.
+        let output_end = if duplicate_pos == 0 {
+            // If the first timestamp of the top node is duplicate. We can simply return the first row
+            // as the heap ensure it is the one with largest sequence.
+            1
+        } else {
+            // We don't know which one has the larger sequence so we use the range before
+            // the duplicate pos.
+            duplicate_pos
+        };
+        Self::maybe_output_batch(
+            top.slice(0, output_end),
+            &mut self.output_batch,
+            self.dedup,
+            &mut self.metrics,
+        )?;
+        top_node.skip_rows(output_end, &mut self.metrics).await?;
+        self.reheap(top_node)
```

Comment:
This is similar to the memtable [merger](https://github.com/GreptimeTeam/greptimedb/blob/175929426a8f06c03a087b1d3b380b9f7af2a474/src/mito2/src/memtable/partition_tree/merger.rs#L113-L122)